File: client/lib/assistant/wav_inspect.dart
/// Shortest WAV we will send to the transcription API.
///
/// This is defensive validation, not a recorder-truncation fix. Android
/// capture must still produce a wall-clock-length recording.
const int minTranscribableWavDurationMs = 500;

/// Structural WAV checks. Never logs or returns audio samples.
class WavInspect {
  const WavInspect({
    required this.validHeader,
    required this.channels,
    required this.sampleRate,
    required this.bitsPerSample,
    required this.dataBytes,
    required this.fileBytes,
  });

  final bool validHeader;
  final int channels;
  final int sampleRate;
  final int bitsPerSample;
  final int dataBytes;
  final int fileBytes;

  int get durationMs {
    final byteRate = sampleRate * channels * (bitsPerSample ~/ 8);
    if (byteRate <= 0) {
      return 0;
    }
    return (dataBytes * 1000) ~/ byteRate;
  }

  Map<String, Object> get debugFields => {
    'wav_header_ok': validHeader,
    'wav_channels': channels,
    'wav_sample_rate': sampleRate,
    'wav_bits': bitsPerSample,
    'wav_data_bytes': dataBytes,
    'file_bytes': fileBytes,
    'duration_ms': durationMs,
  };
}

bool shouldRejectWavForTranscription(WavInspect inspect) {
  return !inspect.validHeader ||
      inspect.durationMs < minTranscribableWavDurationMs;
}

WavInspect? inspectWav(List<int> bytes) {
  if (bytes.length < 44) {
    return null;
  }
  if (!_eq(bytes, 0, 'RIFF') || !_eq(bytes, 8, 'WAVE')) {
    return null;
  }
  var offset = 12;
  var channels = 0;
  var sampleRate = 0;
  var bits = 0;
  var dataBytes = 0;
  var foundFmt = false;
  var foundData = false;
  while (offset + 8 <= bytes.length) {
    final id = String.fromCharCodes(bytes.sublist(offset, offset + 4));
    final size = _u32(bytes, offset + 4);
    final next = offset + 8 + size + (size.isOdd ? 1 : 0);
    if (id == 'fmt ' && size >= 16 && offset + 24 <= bytes.length) {
      foundFmt = true;
      channels = _u16(bytes, offset + 10);
      sampleRate = _u32(bytes, offset + 12);
      bits = _u16(bytes, offset + 22);
    } else if (id == 'data') {
      foundData = true;
      dataBytes = size;
    }
    if (next <= offset) {
      break;
    }
    offset = next;
  }
  return WavInspect(
    validHeader: foundFmt && foundData && channels > 0 && sampleRate > 0,
    channels: channels,
    sampleRate: sampleRate,
    bitsPerSample: bits,
    dataBytes: dataBytes,
    fileBytes: bytes.length,
  );
}

bool _eq(List<int> bytes, int offset, String ascii) {
  for (var i = 0; i < ascii.length; i++) {
    if (bytes[offset + i] != ascii.codeUnitAt(i)) {
      return false;
    }
  }
  return true;
}

int _u16(List<int> bytes, int offset) =>
    bytes[offset] | (bytes[offset + 1] << 8);

int _u32(List<int> bytes, int offset) =>
    bytes[offset] |
    (bytes[offset + 1] << 8) |
    (bytes[offset + 2] << 16) |
    (bytes[offset + 3] << 24);

File: backend/app/assistant/wav_inspect.py
"""Structural WAV checks. Never logs or returns audio samples."""

MIN_TRANSCRIBABLE_WAV_DURATION_MS = 500


def inspect_wav(data: bytes) -> dict | None:
    if len(data) < 44:
        return None
    if data[0:4] != b"RIFF" or data[8:12] != b"WAVE":
        return None
    offset = 12
    channels = 0
    sample_rate = 0
    bits = 0
    data_bytes = 0
    found_fmt = False
    found_data = False
    length = len(data)
    while offset + 8 <= length:
        chunk_id = data[offset : offset + 4]
        size = int.from_bytes(data[offset + 4 : offset + 8], "little")
        next_offset = offset + 8 + size + (size % 2)
        if chunk_id == b"fmt " and size >= 16 and offset + 24 <= length:
            found_fmt = True
            channels = int.from_bytes(data[offset + 10 : offset + 12], "little")
            sample_rate = int.from_bytes(data[offset + 12 : offset + 16], "little")
            bits = int.from_bytes(data[offset + 22 : offset + 24], "little")
        elif chunk_id == b"data":
            found_data = True
            data_bytes = size
        if next_offset <= offset:
            break
        offset = next_offset
    byte_rate = sample_rate * channels * (bits // 8) if bits else 0
    duration_ms = (data_bytes * 1000) // byte_rate if byte_rate else 0
    valid_header = found_fmt and found_data and channels > 0 and sample_rate > 0
    return {
        "valid_header": valid_header,
        "channels": channels,
        "sample_rate": sample_rate,
        "bits_per_sample": bits,
        "data_bytes": data_bytes,
        "file_bytes": length,
        "duration_ms": duration_ms,
    }


def should_reject_wav_for_transcription(inspect: dict) -> bool:
    return (not inspect["valid_header"]) or inspect["duration_ms"] < MIN_TRANSCRIBABLE_WAV_DURATION_MS
